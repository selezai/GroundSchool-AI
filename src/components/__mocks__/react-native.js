const React = require('react');

const View = ({ children, style, testID }) => 
  React.createElement('View', { style, testID }, children);

const Text = ({ children, style, testID }) => 
  React.createElement('Text', { style, testID }, children);

const TextInput = (props) => 
  React.createElement('TextInput', props);

const TouchableOpacity = ({ children, onPress, style, testID }) => 
  React.createElement('TouchableOpacity', { onPress, style, testID }, children);

const StyleSheet = {
  create: styles => styles,
};

module.exports = {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  StyleSheet,
};
