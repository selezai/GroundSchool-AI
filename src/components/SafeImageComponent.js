import React, { useState } from 'react';
import { Image, View, StyleSheet } from 'react-native';
import Logger from '../utils/Logger';

/**
 * SafeImageComponent provides error-resistant image loading
 * to prevent app crashes due to image loading failures
 * 
 * @param {Object} props - Component props
 * @param {any} props.source - Image source (require or uri)
 * @param {Object} props.style - Style for the image
 * @param {string} props.resizeMode - Image resize mode
 * @param {Function} props.onError - Optional callback for error handling
 * @param {Object} props.fallbackSource - Optional fallback image to show on error
 * @param {Object} props.containerStyle - Optional style for the container
 */
const SafeImageComponent = ({
  source,
  style,
  resizeMode = 'contain',
  onError,
  fallbackSource,
  containerStyle,
  ...otherProps
}) => {
  const [hasError, setHasError] = useState(false);
  const [imageLoaded, setImageLoaded] = useState(false);
  
  const handleError = (error) => {
    const errorMessage = error?.nativeEvent?.error || 'Unknown image error';
    Logger.warn(`Image load error: ${errorMessage}`);
    setHasError(true);
    
    if (onError) {
      try {
        onError(error);
      } catch (callbackError) {
        Logger.error('Error in image onError callback', callbackError);
      }
    }
  };
  
  const handleLoad = () => {
    setImageLoaded(true);
  };
  
  // If we have an error and a fallback source, show the fallback
  const imageSource = hasError && fallbackSource ? fallbackSource : source;
  
  // Safely render the image with error handling
  try {
    return (
      <View style={[styles.container, containerStyle]}>
        <Image
          source={imageSource}
          style={[
            styles.image,
            style,
            !imageLoaded && styles.hidden
          ]}
          resizeMode={resizeMode}
          onError={handleError}
          onLoad={handleLoad}
          {...otherProps}
        />
      </View>
    );
  } catch (error) {
    // Catch any synchronous errors during render
    Logger.error('Critical error rendering image', error);
    
    // Return empty view as fallback
    return <View style={[styles.container, containerStyle]} />;
  }
};

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  image: {
    width: '100%',
    height: '100%',
  },
  hidden: {
    opacity: 0,
  },
});

export default SafeImageComponent;
