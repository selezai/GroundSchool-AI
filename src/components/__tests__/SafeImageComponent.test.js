/* global jest, describe, beforeEach, afterEach, it, expect, test */

import React from 'react';
import { Image } from 'react-native';
import renderer, { act } from 'react-test-renderer';
import SafeImageComponent from '../SafeImageComponent';

// Mock Logger
jest.mock('../../utils/Logger', () => ({
  warn: jest.fn(),
  error: jest.fn(),
}));

describe('SafeImageComponent', () => {
  const defaultProps = {
    source: { uri: 'https://example.com/image.jpg' },
    style: { width: 100, height: 100 },
  };

  beforeEach(() => {
    // Clear mock calls between tests
    jest.clearAllMocks();
  });

  it('renders correctly with default props', () => {
    const tree = renderer.create(
      <SafeImageComponent {...defaultProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles image load success', () => {
    let component;
    act(() => {
      component = renderer.create(
        <SafeImageComponent {...defaultProps} />
      );
    });

    const image = component.root.findByType(Image);
    
    act(() => {
      image.props.onLoad();
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('handles image load error', () => {
    const onError = jest.fn();
    const fallbackSource = { uri: 'https://example.com/fallback.jpg' };
    let component;

    act(() => {
      component = renderer.create(
        <SafeImageComponent 
          {...defaultProps}
          onError={onError}
          fallbackSource={fallbackSource}
        />
      );
    });

    const image = component.root.findByType(Image);
    const error = { nativeEvent: { error: 'Failed to load image' } };

    act(() => {
      image.props.onError(error);
    });

    // Should call onError callback
    expect(onError).toHaveBeenCalledWith(error);
    
    // Should use fallback image
    expect(image.props.source).toBe(fallbackSource);
    
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('handles error in onError callback', () => {
    const onError = jest.fn().mockImplementation(() => {
      throw new Error('Callback error');
    });

    let component;
    act(() => {
      component = renderer.create(
        <SafeImageComponent 
          {...defaultProps}
          onError={onError}
        />
      );
    });

    const image = component.root.findByType(Image);
    const error = { nativeEvent: { error: 'Failed to load image' } };

    act(() => {
      image.props.onError(error);
    });

    // Should still render despite callback error
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('applies custom container style', () => {
    const containerStyle = { backgroundColor: 'red' };
    const tree = renderer.create(
      <SafeImageComponent 
        {...defaultProps}
        containerStyle={containerStyle}
      />
    ).toJSON();

    expect(tree.props.style).toContainEqual(containerStyle);
    expect(tree).toMatchSnapshot();
  });

  it('applies custom resize mode', () => {
    const resizeMode = 'cover';
    const component = renderer.create(
      <SafeImageComponent 
        {...defaultProps}
        resizeMode={resizeMode}
      />
    );

    const image = component.root.findByType(Image);
    expect(image.props.resizeMode).toBe(resizeMode);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
