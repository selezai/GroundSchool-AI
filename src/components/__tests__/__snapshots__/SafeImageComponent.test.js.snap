// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeImageComponent applies custom container style 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      {
        "backgroundColor": "red",
      },
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
</View>
`;

exports[`SafeImageComponent applies custom resize mode 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="cover"
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
</View>
`;

exports[`SafeImageComponent handles error in onError callback 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
</View>
`;

exports[`SafeImageComponent handles image load error 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/fallback.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
</View>
`;

exports[`SafeImageComponent handles image load success 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        false,
      ]
    }
  />
</View>
`;

exports[`SafeImageComponent renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      [
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "height": 100,
          "width": 100,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
</View>
`;
